README.md
# [<img src="https://raw.githubusercontent.com/AndyObtiva/glimmer/master/images/glimmer-logo-hi-res.png" height=85 />](https://github.com/AndyObtiva/glimmer) Glimmer DSL for GTK 0.0.9
## Ruby-GNOME Desktop Development GUI Library
[![Gem Version](https://badge.fury.io/rb/glimmer-dsl-gtk.svg)](http://badge.fury.io/rb/glimmer-dsl-gtk)
[![Join the chat at https://gitter.im/AndyObtiva/glimmer](https://badges.gitter.im/AndyObtiva/glimmer.svg)](https://gitter.im/AndyObtiva/glimmer?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[Glimmer](https://github.com/AndyObtiva/glimmer) DSL for [GTK](https://www.gtk.org/) enables building desktop applications with [Ruby-GNOME](https://github.com/ruby-gnome/ruby-gnome) (including [Cairo graphics](#declarative-cairo-graphics)) using [MRI Ruby](https://www.ruby-lang.org).

[GTK](https://www.gtk.org/) (aka GIMP-Toolkit or [incorrectly] GNOME-Toolkit) is the premiere desktop GUI toolkit on [Linux](https://www.gtk.org/docs/installations/linux/) (included in [GNOME flavors of Linux](https://www.gnome.org/getting-gnome/), where [Glimmer DSL for GTK](https://rubygems.org/gems/glimmer-dsl-gtk)'s forte is), which also runs on [Mac](https://www.gtk.org/docs/installations/macos/) ([Quartz GTK+](https://wiki.gnome.org/Projects/GTK/OSX)) and [Windows](https://www.gtk.org/docs/installations/windows).

[Glimmer DSL for GTK](https://rubygems.org/gems/glimmer-dsl-gtk) aims to supercharge productivity and maintainability in developing [Ruby-GNOME](https://github.com/ruby-gnome/ruby-gnome) applications by providing a DSL similar to [Glimmer DSL for SWT](https://github.com/AndyObtiva/glimmer-dsl-swt) having:
- Declarative DSL syntax that visually maps to the GUI widget hierarchy
- Convention over configuration via smart defaults and automation of low-level details
- Requiring the least amount of syntax possible to build GUI
- Custom Keyword support
- Bidirectional Data-Binding to declaratively wire and automatically synchronize GUI with Business Models
- Scaffolding for new custom widgets, apps, and gems
- Native-Executable packaging on Mac, Windows, and Linux.

Hello, World!

```ruby
window {
  title 'Hello, World!'
  
  label('Hello, World!')
}.show
```

Linux | Mac | Windows
------|-----|--------
![hello world screenshot Windows](/screenshots/glimmer-dsl-gtk-linux-hello-world.png) | ![hello world screenshot Mac](/screenshots/glimmer-dsl-gtk-mac-hello-world.png) | ![hello world screenshot Windows](/screenshots/glimmer-dsl-gtk-windows-hello-world.png)

NOTE: Glimmer DSL for GTK is currently in early alpha mode (incomplete proof-of-concept). If you want it developed faster, then [open an issue report](https://github.com/AndyObtiva/glimmer-dsl-gtk/issues/new). I have completed some GitHub project features much faster before due to [issue reports](https://github.com/AndyObtiva/glimmer-dsl-gtk/issues) and [pull requests](https://github.com/AndyObtiva/glimmer-dsl-gtk/pulls). Please help make better by contributing, adopting for small or low risk projects, and providing feedback. It is still an early alpha, so the more feedback and issues you report the better.

**[Glimmer](https://rubygems.org/gems/glimmer) DSL Comparison Table:**
DSL | Platforms | Native? | Vector Graphics? | Pros | Cons | Prereqs
----|-----------|---------|------------------|------|------|--------
[Glimmer DSL for SWT (JRuby Desktop Development GUI Framework)](https://github.com/AndyObtiva/glimmer-dsl-swt) | Mac / Windows / Linux | Yes | Yes (Canvas Shape DSL) | Very Mature / Scaffolding / Native Executable Packaging / Custom Widgets | Slow JRuby Startup Time / Heavy Memory Footprint | Java / JRuby
[Glimmer DSL for Opal (Pure Ruby Web GUI and Auto-Webifier of Desktop Apps)](https://github.com/AndyObtiva/glimmer-dsl-opal) | All Web Browsers | No | Yes (Canvas Shape DSL) | Simpler than All JavaScript Technologies / Auto-Webify Desktop Apps | Setup Process / Only Rails 5 Support for Now | Rails
[Glimmer DSL for LibUI (Prerequisite-Free Ruby Desktop Development GUI Library)](https://github.com/AndyObtiva/glimmer-dsl-libui) | Mac / Windows / Linux | Yes | Yes (Area API) | Fast Startup Time / Light Memory Footprint | LibUI is an Incomplete Mid-Alpha Only | None Other Than MRI Ruby
[Glimmer DSL for Tk (MRI Ruby Desktop Development GUI Library)](https://github.com/AndyObtiva/glimmer-dsl-tk) | Mac / Windows / Linux | Some Native-Themed Widgets (Not Truly Native) | Yes (Canvas) | Fast Startup Time / Light Memory Footprint | Widgets Do Not Look Truly Native, Espcially on Linux | ActiveTcl / MRI Ruby
[Glimmer DSL for GTK (Ruby-GNOME Desktop Development GUI Library)](https://github.com/AndyObtiva/glimmer-dsl-gtk) | Mac / Windows / Linux | Only on Linux | Yes (Cairo) | Complete Access to GNOME Features on Linux (Forte) | Not Native on Mac and Windows | None Other Than MRI Ruby on Linux / Brew Packages on Mac / MSYS & MING Toolchains on Windows / MRI Ruby
[Glimmer DSL for FX (FOX Toolkit Ruby Desktop Development GUI Library)](https://github.com/AndyObtiva/glimmer-dsl-fx) | Mac (requires XQuartz) / Windows / Linux | No | Yes (Canvas) | No Prerequisites on Windows (Forte Since Binaries Are Included Out of The Box) | Widgets Do Not Look Native / Mac Usage Obtrusively Starts XQuartz | None Other Than MRI Ruby on Windows / XQuarts on Mac / MRI Ruby
[Glimmer DSL for JFX (JRuby JavaFX Desktop Development GUI Library)](https://github.com/AndyObtiva/glimmer-dsl-jfx) | Mac / Windows / Linux | No | Yes (javafx.scene.shape and javafx.scene.canvas) | Rich in Custom Widgets | Slow JRuby Startup Time / Heavy Memory Footprint / Widgets Do Not Look Native | Java / JRuby / JavaFX SDK
[Glimmer DSL for Swing (JRuby Swing Desktop Development GUI Library)](https://github.com/AndyObtiva/glimmer-dsl-swing) | Mac / Windows / Linux | No | Yes (Java2D) | Very Mature | Slow JRuby Startup Time / Heavy Memory Footprint / Widgets Do Not Look Native | Java / JRuby
[Glimmer DSL for XML (& HTML)](https://github.com/AndyObtiva/glimmer-dsl-xml) | All Web Browsers | No | Yes (SVG) | Programmable / Lighter-weight Than Actual XML | XML Elements Are Sometimes Not Well-Named (Many Types of Input) | None
[Glimmer DSL for CSS](https://github.com/AndyObtiva/glimmer-dsl-css) | All Web Browsers | No | Yes | Programmable | CSS Is Over-Engineered / Too Many Features To Learn | None

## Prerequisites

### Linux

GNOME-based Linux installations do not need any prerequisites since they have GTK built-in, so you can skip to [Setup](#setup) right away.

Still, if you run into issues installing the GTK gem in [Setup](#setup), you may try these commands first:
```
sudo apt-get install libgtk-3-dev
```

Or run this command if all else fails:
```
sudo apt install -y -V libgirepository1.0-dev
```

### Mac

On the Mac, make sure to:
- Have [Homebrew](https://brew.sh/) installed
- Run this [Homebrew](https://brew.sh/) command to have GTK display GUI icons: `brew install adwaita-icon-theme`
- (Optional) You can upgrade your GTK3/GTK4/GTK+ by running: `brew install gtk+3` / `brew install gtk+4` / `brew install gtk+`

### Windows

Make sure to install [Ruby](https://www.ruby-lang.org) with MSYS & MING toolchains from [RubyInstaller](https://rubyinstaller.org/downloads/)

## Setup

### Option 1: Install

Run this command to install directly:
```
gem install glimmer-dsl-gtk
```

### Option 2: Bundler

Add the following to `Gemfile`:
```
gem 'glimmer-dsl-gtk', '~> 0.0.9'
```

And, then run:
```
bundle
```

## Usage

Require the library and mixin the `Glimmer` module to utilize the Glimmer GUI DSL for GTK:

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Demo'
  
  on(:destroy) do
    puts 'Bye Bye'
    ::Gtk.main_quit
  end
}.show
```

For actual application development outside of simple demos, mixin the `Glimmer` module into a custom application class instead:

```ruby
require 'glimmer-dsl-gtk'

class SomeGlimmerApplication
  include Glimmer
  
  def launch
    application('org.glimmer.hello-application', :flags_none) {
      on(:activate) do |app|
        application_window(app) {
          title 'Actual Application'
        }.present
      end
    }.run
  end
end

SomeGlimmerApplication.new.launch
```

### Glimmer GUI DSL

- Keywords: All GTK widgets are supported via lowercase underscored names accepting their constructor args (e.g. `application_window(app)` for `Gtk::ApplicationWindow.new(app)`). Keywords can be nested under other keywords to represent the true hierarchy of nested widgets on the screen (e.g. `window { label('Hello') }` is a `label` nested under a `window`). Note that widget objects returned are proxies of the GTK widget counterparts. This shields consumers of GTK from its lower-level details via composition (Proxy Design Pattern). To access lower-level GTK widget, simply call `#gtk` method on widget proxy object (e.g. `@w = window {...}; @w.gtk # Gtk::Window widget object`).
- Content: widget keywords can have a block of content that could contain nested widget keywords, properties, and signals. The block can optionally receive one argument representing the widget (e.g. `window {|w| ... }`):
  - Properties: All GTK widget properties can be set via lowercase underscored names (without the 'set_' prefix) nested under widget keywords (e.g. `window {title 'Hello, World'}` sets `title` property of `window`)
  - Signals: All GTK signals can be wired with `on(signal) { ... }` syntax (e.g. `on(:activate) { do_something }`)

### MVC Observer Pattern

In Smalltalk-MVC ([Model View Controller](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) Architectural Pattern), the View is an active View that observes the Model for changes and updates itself.

![MVC](http://3.bp.blogspot.com/-4eW59Ao0ess/ToiBzAiYdZI/AAAAAAAAAOg/SiYa6XHwBFE/s320/Screen+shot+2011-10-02+at+10.22.11+AM.png)

This can be achieved with the Glimmer GUI DSL using the `observe` keyword, which takes a model (any object, including `self`) and attribute Symbol or String expression (e.g. `:count` or `'address.street'`).

The model is automatically enhanced as an `Glimmer::DataBinding::ObservableModel` / `Glimmer::DataBinding::ObservableHash` / `Glimmer::DataBinding::ObservableArray` depending on its type to support notifying observers of attribute changes (when performed using the attribute writer, which automatically calls added method `notify_observers(attribute)`)

Note that it is usually recommended to observe external model objects (not `self`), but `self` is OK in very simple cases or presentation-related attributes only.

### Declarative Cairo Graphics

[Cairo](https://www.cairographics.org/) is the engine behind drawing arbitrary 2D geometric shapes in [GTK](https://www.gtk.org/).

In [Glimmer DSL for GTK](https://rubygems.org/gems/glimmer-dsl-gtk), you can draw Cairo shapes declaratively in a way similar to how SVG works, but using one language; Ruby, thus being able to utilize Ruby logic (e.g. if statement or each loop) with it effortlessly when needed. Declarative syntax also yields less code that is simpler, not dependent on ordering of nested properties, and more understandable/maintainable.

Below is a quick tutorial consisting of samples inspired and ported from [Mohit Sindhwani's blog post "Cairo with Ruby - Samples using RCairo"](https://notepad.onghu.com/2021/cairo-samples-in-ruby/).

#### Arc

[samples/cairo/arc.rb](/samples/cairo/arc.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Arc'
  default_size 256, 256
  
  drawing_area {
    # Surface Paint
    paint 242.25, 242.25, 242.25
    
    # Set up the parameters
    xc = 128.0
    yc = 128.0
    radius = 100.0
    angle1 = 45.0  * (Math::PI/180.0) # angles are specified
    angle2 = 180.0  * (Math::PI/180.0) # in radians
    
    # The main arc
    arc(xc, yc, radius, angle1, angle2) {
      stroke 0, 0, 0
      line_width 10
    }
    
    # Draw helping lines
    
    # First, the circle at the centre
    arc(xc, yc, 10.0, 0, 2*Math::PI) {
      fill 255, 51, 51, 0.6
    }
    
    # Then, the lines reaching out
    path {
      arc xc, yc, radius, angle1, angle1
      line_to xc, yc
      arc xc, yc, radius, angle2, angle2
      line_to xc, yc
      
      stroke 255, 51, 51, 0.6
      line_width 6
    }
  }
}.show
```

![Arc](/screenshots/glimmer-dsl-gtk-mac-cairo-arc.png)

#### Arc Negative

[samples/cairo/arc_negative.rb](/samples/cairo/arc_negative.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Arc Negative'
  default_size 256, 256
  
  drawing_area {
    # Surface Paint
    paint 255, 255, 255
    
    # Set up the parameters
    xc = 128.0
    yc = 128.0
    radius = 100.0
    angle1 = 45.0  * (Math::PI/180.0) # angles are specified
    angle2 = 180.0  * (Math::PI/180.0) # in radians
    
    # The main negative arc
    arc_negative(xc, yc, radius, angle1, angle2) {
      stroke 0, 0, 0
      line_width 10
    }
    
    # Draw helping lines
    
    # First, the circle at the centre
    arc(xc, yc, 10.0, 0, 2*Math::PI) {
      fill 255, 51, 51, 0.6
    }
    
    # Then, the lines reaching out
    path {
      arc(xc, yc, radius, angle1, angle1)
      line_to(xc, yc)
      arc(xc, yc, radius, angle2, angle2)
      line_to(xc, yc)
      
      stroke 255, 51, 51, 0.6
      line_width 6
    }
  }
}.show
```

![Arc Negative](/screenshots/glimmer-dsl-gtk-mac-cairo-arc-negative.png)

#### Clip

[samples/cairo/clip.rb](/samples/cairo/clip.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Clip'
  default_size 256, 256
  
  drawing_area {
    # Surface Paint
    paint 255, 255, 255
    
    # Designate arc as the clipping area
    arc(128.0, 128.0, 76.8, 0, 2 * Math::PI) {
      clip true
    }
    
    # Rectangle will get clipped by arc
    rectangle(0, 0, 256, 256) {
      fill 0, 0, 0
    }
    
    # Path will get clipped by arc
    path {
      move_to 0, 0
      line_to 256, 256
      move_to 256, 0
      line_to 0, 256
      
      stroke 0, 255, 0
      line_width 10
    }
  }
}.show
```

![Clip](/screenshots/glimmer-dsl-gtk-mac-cairo-clip.png)

#### Clip Image

[samples/cairo/clip_image.rb](/samples/cairo/clip_image.rb)

```ruby
require 'glimmer-dsl-gtk'
require 'net/http'

image_content = Net::HTTP.get(URI('https://raw.githubusercontent.com/AndyObtiva/glimmer-dsl-gtk/master/images/breaking-blue-wave.png'))
image_file = File.join(Dir.home, 'breaking-blue-wave.png')
File.write(image_file, image_content)

include Glimmer

window {
  title 'Clip Image'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25

    arc(128.0, 128.0, 76.8, 0, 2 * Math::PI) {
      clip true # designate arc as the clipping area
    }

    rectangle(0, 0, 256, 256) {
      # Source image is from:
      # - https://www.publicdomainpictures.net/en/view-image.php?image=7683&picture=breaking-blue-wave
      # Converted to PNG before using it
      image = Cairo::ImageSurface.from_png(image_file)
      w = image.width
      h = image.height
  
      scale 256.0/w, 256.0/h, exclude: :shape # applies scale to fill source image only
      fill image, 0, 0
    }
  }
}.show
```

![Clip Image](/screenshots/glimmer-dsl-gtk-mac-cairo-clip-image.png)

#### Curve to

[samples/cairo/curve_to.rb](/samples/cairo/curve_to.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Curve to'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    x=25.6
    y=128.0
    x1=102.4
    y1=230.4
    x2=153.6
    y2=25.6
    x3=230.4
    y3=128.0
    
    path {
      move_to x, y
      curve_to x1, y1, x2, y2, x3, y3
      
      line_width 10
      stroke 0, 0, 0
    }
    
    path {
      move_to x,y
      line_to x1,y1
      move_to x2,y2
      line_to x3,y3
      
      line_width 6
      stroke 255, 51, 51, 0.6
    }
  }
}.show
```

![Curve to](/screenshots/glimmer-dsl-gtk-mac-cairo-curve-to.png)

#### Dashes

[samples/cairo/dashes.rb](/samples/cairo/dashes.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Dashes'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25

    dashes = [ 50.0, # ink
                       10.0,  # skip
                       10.0,  # ink
                       10.0   # skip
            ]
    offset = -50.0
    
    path {
      move_to 128.0, 25.6
      line_to 230.4, 230.4
      rel_line_to -102.4, 0.0
      curve_to 51.2, 230.4, 51.2, 128.0, 128.0, 128.0
      
      line_width 10
      dash dashes, offset
      stroke 0, 0, 0
    }
  }
}.show
```

![Dashes](/screenshots/glimmer-dsl-gtk-mac-cairo-dashes.png)

#### Fill and Stroke 2

(note: there is no Fill and Stroke 1; this was adopted from [Mohit's blog post](https://notepad.onghu.com/2021/cairo-samples-in-ruby/), which only mentioned Fill and Stroke 2)

[samples/cairo/fill_and_stroke2.rb](/samples/cairo/fill_and_stroke2.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Fill and Stroke 2'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    path {
      move_to 128.0, 25.6
      line_to 230.4, 230.4
      rel_line_to -102.4, 0.0
      curve_to 51.2, 230.4, 51.2, 128.0, 128.0, 128.0
      close_path

      fill 0, 0, 255
      stroke 0, 0, 0
      line_width 10
    }
    
    path {
      move_to 64.0, 25.6
      rel_line_to 51.2, 51.2
      rel_line_to -51.2, 51.2
      rel_line_to -51.2, -51.2
      close_path

      fill 0, 0, 255
      stroke 0, 0, 0
      line_width 10
    }
  }
}.show
```

![Fill and Stroke 2](/screenshots/glimmer-dsl-gtk-mac-cairo-fill-and-stroke2.png)

#### Fill Style

[samples/cairo/fill_style.rb](/samples/cairo/fill_style.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Fill Style'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    path {
      rectangle 12, 12, 232, 70
      path { # sub-path
        arc 64, 64, 40, 0, 2*Math::PI
      }
      path { # sub-path
        arc_negative 192, 64, 40, 0, -2*Math::PI
      }
      
      fill_rule Cairo::FILL_RULE_EVEN_ODD
      line_width 6
      fill 0, 178.5, 0
      stroke 0, 0, 0
    }
    
    path {
      rectangle 12, 12, 232, 70
      path { # sub-path
        arc 64, 64, 40, 0, 2*Math::PI
      }
      path { # sub-path
        arc_negative 192, 64, 40, 0, -2*Math::PI
      }
      
      translate 0, 128
      fill_rule Cairo::FILL_RULE_WINDING
      line_width 6
      fill 0, 0, 229.5
      stroke 0, 0, 0
    }
  }
}.show
```

![Fill Style](/screenshots/glimmer-dsl-gtk-mac-cairo-fill-style.png)

#### Gradient

[samples/cairo/gradient.rb](/samples/cairo/gradient.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Gradient'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
        
    # Create the Linear Pattern
    rectangle(0, 0, 256, 256) {
      pat = Cairo::LinearPattern.new(0.0, 0.0,  0.0, 256.0)
      pat.add_color_stop_rgba(1, 0, 0, 0, 1)
      pat.add_color_stop_rgba(0, 1, 1, 1, 1)
      
      fill pat
    }
    
    # Create the radial pattern
    arc(128.0, 128.0, 76.8, 0, 2 * Math::PI) {
      pat = Cairo::RadialPattern.new(115.2, 102.4, 25.6,
                                         102.4,  102.4, 128.0)
      pat.add_color_stop_rgba(0, 1, 1, 1, 1)
      pat.add_color_stop_rgba(1, 0, 0, 0, 1)
      
      fill pat
    }
  }
}.show
```

![Gradient](/screenshots/glimmer-dsl-gtk-mac-cairo-gradient.png)

#### Image

[samples/cairo/image.rb](/samples/cairo/image.rb)

```ruby
require 'glimmer-dsl-gtk'
require 'net/http'

image_content = Net::HTTP.get(URI('https://raw.githubusercontent.com/AndyObtiva/glimmer-dsl-gtk/master/images/breaking-blue-wave.png'))
image_file = File.join(Dir.home, 'breaking-blue-wave.png')
File.write(image_file, image_content)

include Glimmer

window {
  title 'Image'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
        
    image = Cairo::ImageSurface.from_png(image_file)
    w = image.width
    h = image.height
    
    translate 128.0, 128.0
    rotate 45*Math::PI/180
    scale 256.0/w, 256.0/h
    translate -0.5*w, -0.5*h
    
    paint image, 0, 0
  }
}.show
```

![Image](/screenshots/glimmer-dsl-gtk-mac-cairo-image.png)

#### Image Gradient

[samples/cairo/image_gradient.rb](/samples/cairo/image_gradient.rb)

```ruby
require 'glimmer-dsl-gtk'
require 'net/http'

image_content = Net::HTTP.get(URI('https://raw.githubusercontent.com/AndyObtiva/glimmer-dsl-gtk/master/images/breaking-blue-wave.png'))
image_file = File.join(Dir.home, 'breaking-blue-wave.png')
File.write(image_file, image_content)

include Glimmer

window {
  title 'Image Gradient'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    image = Cairo::ImageSurface.from_png(image_file)
    w = image.width
    h = image.height
    
    # Load the image as a surface pattern
    pattern = Cairo::SurfacePattern.new(image)
    pattern.extend = Cairo::EXTEND_REPEAT
    
    # Set up the scale matrix
    pattern.matrix = Cairo::Matrix.scale(w/256.0 * 5.0, h/256.0 * 5.0)
    
    rectangle(0, 0, 256, 256) {
      translate 128.0, 128.0
      rotate Math::PI / 4
      scale 1/Math.sqrt(2), 1/Math.sqrt(2)
      translate -128.0, -128.0
        
      fill pattern
    }
  }
}.show
```

![Image Gradient](/screenshots/glimmer-dsl-gtk-mac-cairo-image-gradient.png)

#### Multi Segment Caps

[samples/cairo/multi_segment_caps.rb](/samples/cairo/multi_segment_caps.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Multi Segment Caps'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    path {
      move_to 50.0, 75.0
      line_to 200.0, 75.0
      
      move_to 50.0, 125.0
      line_to 200.0, 125.0
      
      move_to 50.0, 175.0
      line_to 200.0, 175.0
      
      line_width 30
      line_cap Cairo::LINE_CAP_ROUND
      stroke 0, 0, 0
    }
  }
}.show
```

![Multi Segment Caps](/screenshots/glimmer-dsl-gtk-mac-cairo-multi-segment-caps.png)

#### Rounded Rectangle

[samples/cairo/rounded_rectangle.rb](/samples/cairo/rounded_rectangle.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Rounded Rectangle'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    path {
      rounded_rectangle(25.6, 25.6, 204.8, 204.8, 20)
    
      fill 127.5, 127.5, 255
      line_width 10.0
      stroke 127.5, 0, 0, 0.5
    }
  }
}.show
```

![Rounded Rectangle](/screenshots/glimmer-dsl-gtk-mac-cairo-rounded-rectangle.png)

#### Set line cap

[samples/cairo/set_line_cap.rb](/samples/cairo/set_line_cap.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Set line cap'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    # The main code
    path {
      move_to 64.0, 50.0
      line_to 64.0, 200.0
      
      line_cap Cairo::LINE_CAP_BUTT #  default
      line_width 30
      stroke 0, 0, 0
    }
    
    path {
      move_to 128.0, 50.0
      line_to 128.0, 200.0
      
      line_cap Cairo::LINE_CAP_ROUND
      line_width 30
      stroke 0, 0, 0
    }
    
    path {
      move_to 192.0, 50.0
      line_to 192.0, 200.0
      
      line_cap Cairo::LINE_CAP_SQUARE
      line_width 30
      stroke 0, 0, 0
    }
    
    #  draw helping lines */
    path {
      move_to 64.0, 50.0
      line_to 64.0, 200.0
      move_to 128.0, 50.0
      line_to 128.0, 200.0
      move_to 192.0, 50.0
      line_to 192.0, 200.0
      
      line_width 2.56
      stroke 255, 51, 51
    }
  }
}.show
```

![Set line cap](/screenshots/glimmer-dsl-gtk-mac-cairo-set-line-cap.png)

#### Set line join

[samples/cairo/set_line_join.rb](/samples/cairo/set_line_join.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Set line join'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    # The main code
    path {
      move_to 76.8, 84.48
      rel_line_to 51.2, -51.2
      rel_line_to 51.2, 51.2
      
      line_join Cairo::LINE_JOIN_MITER # default
      line_width 40.96
      stroke 0, 0, 0
    }
    
    path {
      move_to 76.8, 161.28
      rel_line_to 51.2, -51.2
      rel_line_to 51.2, 51.2
      
      line_join Cairo::LINE_JOIN_BEVEL
      line_width 40.96
      stroke 0, 0, 0
    }
    
    path {
      move_to 76.8, 238.08
      rel_line_to 51.2, -51.2
      rel_line_to 51.2, 51.2
      
      line_join Cairo::LINE_JOIN_ROUND
      line_width 40.96
      stroke 0, 0, 0
    }
  }
}.show
```

![Set line join](/screenshots/glimmer-dsl-gtk-mac-cairo-set-line-join.png)

#### Text

[samples/cairo/text.rb](/samples/cairo/text.rb)

```ruby
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Text'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    font_family = OS.linux? ? 'Sans' : (OS.mac? ? 'Helvetica' : 'Arial')
    
    # The main code
    path {
      move_to 10.0, 135.0
      show_text 'Hello'
      
      font_face font_family, Cairo::FONT_SLANT_NORMAL, Cairo::FONT_WEIGHT_BOLD
      font_size 90.0
      line_width 2.56
      fill 0, 0, 0
      stroke 0, 0, 0
    }
    
    path {
      move_to 70.0, 165.0
      text_path 'void'
      
      font_face font_family, Cairo::FONT_SLANT_NORMAL, Cairo::FONT_WEIGHT_BOLD
      font_size 90.0
      line_width 2.56
      fill 127.5, 127.5, 255
      stroke 0, 0, 0
    }
    
    # draw helping lines
    path {
      arc 10.0, 135.0, 5.12, 0, 2*Math::PI
      close_path
      arc 70.0, 165.0, 5.12, 0, 2*Math::PI
      
      fill 255, 51, 51, 0.6
    }
  }
}.show
```


CHANGELOG.md
# Change Log

## 0.0.9

- Add the text "Next" before the Next Tetromino Preview area
- Tetris "Show Next Block Preview" menu item under "View" menu
- Tetris "Speed" menu
- Fix Tetris Help menu (was called Options by mistake)
- Fix issue with Tetris game restarting immediately on game over
- Fix issue with Tetris not accelerating upon level ups
- Fix warning message: glimmer-dsl-gtk-0.0.8/lib/glimmer/gtk/widget_proxy.rb:192:in `new': 'Gtk::MessageDialog#initialize(parent, flags, type, buttons_type, message)' style has been deprecated. Use 'Gtk::MessageDialog#initialize(:parent => nil, :flags => 0, :type => :info, :buttons => :ok, :message => "")' style.
- Fix warning message: Attempting to add a widget with type GtkMenu to a container of type GtkMenuItem, but the widget is already inside a container of type GtkWindow, please remove the widget from its existing container first.

## 0.0.8

- Download `images/breaking-blue-wave.png` from web in Cairo examples that use it (`image.rb`, `clip_image.rb`, `image_gradient.rb`) to avoid bloating Ruby gem

## 0.0.7

- samples/cairo/text.rb
- Support for `path` nested `show_text`, `text_path`, `show_glyphs`, `glyph_path`, `new_sub_path` operations

## 0.0.6

- samples/cairo/clip_image.rb
- samples/cairo/curve_to.rb
- samples/cairo/dashes.rb
- samples/cairo/fill_and_stroke2.rb
- samples/cairo/fill_style.rb
- samples/cairo/gradient.rb
- samples/cairo/image.rb
- samples/cairo/image_gradient.rb
- samples/cairo/multi_segment_caps.rb
- samples/cairo/rounded_rectangle.rb
- samples/cairo/set_line_cap.rb
- samples/cairo/set_line_join.rb
- Supported nested sub-paths (nesting `path` within another `path`)
- Support declarative `fill`/`paint` of `image, x, y` in cairo graphics
- Support declarative transforms on cairo graphics shapes: `translate`, `scale`, `rotate`
- Support applying transform on `drawing_area` `paint`
- Support for `path` nested `rectangle`, `rounded_rectangle`, and `arc_negative` operations

## 0.0.5

- Support `drawing_area#paint(red, green, blue)` operation to set the initial `drawing_area` paint color base
- Support `arc` as an operation inside `path` in Cairo declarative shape syntax
- Support `arc_negative` cairo graphics shape
- Support cairo graphics shape `clip`

## 0.0.4

- Tetris Menu Bar
- Tetris Score Board
- Tetris Next Tetromino Preview

## 0.0.3

- samples/elaborate/tetris.rb (basic Tetris implementation)
- `observe` keyword for observing Model attributes to update the View
- Ensure `fill`/`stroke` properties in `drawin_area` shapes accept 255-based rgb colors (or rgba with 1-based transparency)
- Support setting all cairo shape attributes (e.g. `line_width`, `line_cap`, etc...) with declarative syntax (whichever is specified)
- Upgrade to glimmer 2.6.0

## 0.0.2

- Support re-opening any widget proxy content with `#content {}` block to add more nested widgets inside
- Draw `drawing_area` Cairo shapes declaratively with keywords: `rectangle`, `rounded_rectangle`, `arc`, `triangle`, `circle`, `polygon`, `polyline`, `path`
- samples/hello/hello_drawing_area.rb (declarative cairo drawing)
- samples/hello/hello_drawing_area_manual.rb (use cairo manually with imperative syntax)
- samples/elaborate/widget_gallery.rb

## 0.0.1

- Initial version of Glimmer DSL for GTK
- General widget support with keyword arguments (e.g. `window(:toplevel).show`)
- General widget property support (e.g. `title 'Hello, Application!'`)
- General signal connect support via nested `on` keyword (e.g. `on(:activate) { ... }`)
- Have `window` automatically call `::Gtk::main_quit` upon hitting `window` close (destroy) button
- Automatically set `window` `default_size` to `190, 150` by default
- Automatically set `message_dialog` `default_width` to `190` by default
- Automatically set `application` first arg to `'org.glimmer-dsl-gtk.application'` and second arg to `:flags_none` when not supplied
- Automatically set `box` orientation to `:vertical` if not specified
- Hello, World! sample
- Hello, Application! sample
- Hello, Button! sample


samples/cairo/text.rb
require 'glimmer-dsl-gtk'

include Glimmer

window {
  title 'Text'
  default_size 256, 256
  
  drawing_area {
    paint 242.25, 242.25, 242.25
    
    font_family = OS.linux? ? 'Sans' : (OS.mac? ? 'Helvetica' : 'Arial')
    
    # The main code
    path {
      move_to 10.0, 135.0
      show_text 'Hello'
      
      font_face font_family, Cairo::FONT_SLANT_NORMAL, Cairo::FONT_WEIGHT_BOLD
      font_size 90.0
      line_width 2.56
      fill 0, 0, 0
      stroke 0, 0, 0
    }
    
    path {
      move_to 70.0, 165.0
      text_path 'void'
      
      font_face font_family, Cairo::FONT_SLANT_NORMAL, Cairo::FONT_WEIGHT_BOLD
      font_size 90.0
      line_width 2.56
      fill 127.5, 127.5, 255
      stroke 0, 0, 0
    }
    
    # draw helping lines
    path {
      arc 10.0, 135.0, 5.12, 0, 2*Math::PI
      close_path
      arc 70.0, 165.0, 5.12, 0, 2*Math::PI
      
      fill 255, 51, 51, 0.6
    }
  }
}.show
